Assets/MixedRealityToolkit.SDK/Experimental/Features/UX/WindowsKeyboard/MixedRealityKeyboard.cs

```diff
@@ -0,0 +1,196 @@
+                    text = value;
+                    TextChanged?.Invoke(new KeyboardTextData(oldValue, value));
+                }
+            }
+        }
+        
+        private KeyboardState State
+        {
+            get
+            {
+                return state;
+            }
+
+            set
+            {
+                if (state != value)
+                {
+                    state = value;
+                    if (state == KeyboardState.showing)
+                    {
+                        KeyboardShown.Invoke();
+                    }
+                    else if (state == KeyboardState.hidden)
+                    {
+                        KeyboardHidden.Invoke();
+                    }
+                }
+            }
+        }
+        #endregion properties
+
+        #region Private fields
+        private TouchScreenKeyboard keyboard;
+
+#if !UNITY_EDITOR && UNITY_WSA
+    private InputPane inputPane = null;
+#endif //!UNITY_EDITOR && UNITY_WSA
+
+        private KeyboardState state = KeyboardState.hidden;
+        #endregion private fields
+
+        #region Private enums
+        private enum KeyboardState
+        {
+            hidden,
+            showing,
+            hiding
+        }
+        #endregion Private enums
+
+        #region Unity functions
+        private void Start()
+        {
+#if !UNITY_EDITOR && UNITY_WSA
+        UnityEngine.WSA.Application.InvokeOnUIThread(() =>
+        {
+            inputPane = InputPane.GetForCurrentView();
+            inputPane.Hiding += (inputPane, args) => OnKeyboardHiding();
+            inputPane.Showing += (inputPane, args) => OnKeyboardShowing();
+        }, false);
+#endif //!UNITY_EDITOR && UNITY_WSA
+        }
+
+        private void Update()
+        {
+            switch (State)
+            {
+                case KeyboardState.showing:
+                    CommitText();
+                    break;
+
+                case KeyboardState.hiding:
+                    ClearText();
+                    State = KeyboardState.hidden;
+                    break;
+
+                case KeyboardState.hidden:
+                default:
+                    break;
+            }
+        }
+        #endregion unity functions
+
+        public void HideKeyboard()
+        {
+            ClearText();
+            State = KeyboardState.hidden;
+
+#if !UNITY_EDITOR && UNITY_WSA
+        UnityEngine.WSA.Application.InvokeOnUIThread(() => inputPane?.TryHide(), false);
+#endif //!UNITY_EDITOR && UNITY_WSA
+        }
+
+        public void ShowKeyboard()
+        {
+            if (state == KeyboardState.showing)
+            {
+                Debug.Log($"MixedRealityKeyboard.ShowKeyboard called but keyboard already visible");
+                return;
+            }
+
+            State = KeyboardState.showing;
+
+            if (keyboard != null)
+            {
+                keyboard.text = string.Empty;
+#if !UNITY_EDITOR && UNITY_WSA
```

Comment:
>if !UNITY_EDITOR && UNITY_WSA [](start = 1, length = 29)

Should we have an else clause that will handle the iPhone/Android cases (i.e. I assume Open still works on those)

[https://docs.unity3d.com/ScriptReference/TouchScreenKeyboard.html](https://docs.unity3d.com/ScriptReference/TouchScreenKeyboard.html)

Reply:
Or wait, nevermind, this class explicitly calls out that this is WMR only as it works around a bug that is specific to those systems.

---
In reply to: [296406649](https://github.com/microsoft/MixedRealityToolkit-Unity/pull/5018#discussion_r296406649) [](ancestors = 296406649)